osu.Game/Database/BackgroundDataStoreProcessor.cs

```diff
@@ -86,6 +92,69 @@ protected override void LoadComplete()
             });
         }
 
+        /// <summary>
+        /// Go through every beatmap and calculate the audio normalization values if they are missing.
+        /// </summary>
+        private void processAudioNormalization()
+        {
+            int beatmapsCount = 0;
+            int beatmapsToProcess = 0;
+
+            realmAccess.Run(r =>
+            {
+                foreach (BeatmapSetInfo beatmapSetInfo in r.All<BeatmapSetInfo>().Where(b => !b.DeletePending))
+                {
+                    beatmapsCount += beatmapSetInfo.Beatmaps.Count;
+                    beatmapsToProcess += beatmapSetInfo.Beatmaps.Count(b => b.AudioNormalization == null);
+                }
+            });
+
+            Logger.Log("beatmaps to process: " + beatmapsToProcess);
+            if (beatmapsToProcess == 0)
+            {
+                Logger.Log("No audio normalization processing required.");
+                return;
+            }
+
+            var notification = showProgressNotification(beatmapsCount, "Verifying loudness level for beatmaps", "all loudness levels have been verified");
+            int processedCount = 0;
+
+            realmAccess.Run(r =>
+            {
+                r.Refresh();
+
+                foreach (BeatmapSetInfo beatmapSetInfo in r.All<BeatmapSetInfo>().Where(b => !b.DeletePending))
+                {
+                    foreach (BeatmapInfo beatmapInfo in beatmapSetInfo.Beatmaps)
+                    {
+                        if (notification?.State == ProgressNotificationState.Cancelled)
+                            break;
+
+                        updateNotificationProgress(notification, processedCount, beatmapsCount);
+                        processedCount++;
+
+                        sleepIfRequired();
+                        if (beatmapInfo.AudioNormalization != null) continue;
+
+                        realmAccess.Write(realm =>
+                        {
+                            AudioNormalization audioNormalization = new AudioNormalization(beatmapInfo, beatmapSetInfo, new RealmFileStore(realmAccess, storage!));
+                            beatmapInfo.AudioNormalization = audioNormalization;
+                            realm.Add(audioNormalization.PopulateSet(beatmapInfo, beatmapSetInfo), true);
```

Comment:
I don't think this is required? I wouldn't be surprised if this is your memory leak.

Reply:
Referencing `BeatmapImporter`, you can see that it does `beatmapSet = audioNormalization.PopulateSet()` after setting the audionormalization for the beatmapinfo. That same thing can't be done here, because you can't modify an iterator in a foreach while iterating.

I could try a for loop and instead go with `BeatmapImporter`'s approach, but I am certain that removing `PopulateSet` here would not be wise.

The entire reason that thing exists is because the loops work as so:
- Go through each beatmapset
- For each beatmapset, go through its difficulties
- If the first difficulty has the value of null, then calculate the loudness values for it
- Now check the other difficulties to see if they have the same audio file, if so, just copy over the loudness values from the first difficulty.
- If a difficulty has a different audio file, copy over the loudness values for the difficulties with a similar audio file. Leaving the ones with different audio files as null
- (new loop) Go to the 2nd difficulty, if its null, then continue until we find a difficulty that is still null
- For the null difficulties, we repeat this process, calculate the loudness values, copy them to difficulties with similar audio files, then repeat.
- Finally, after having gone through all of the difficulties, go to the next beatmapset and repeat

If we don't reassign beatmapset to the return value of `PopulateSet`, the loudness values will not be copied over to similar audio files, resulting in *every* difficulty being processed even though their audio files are the same. Adding to realm is a workaround to reassign beatmapset.